File: matlab/Pipeline/CBASS_L2_PlotEmbed.py
# CBASS_L2_PLotEmbed.py;
'''
This function formats and plots the embedding of the data specified by variable 'experiment' in the temporary directory.
The current method makes use of the scprep (https://scprep.readthedocs.io/en/stable/index.html), which should be automatically installed 
when replicating our conda environment (https://github.com/ahof1704/gamma_bouts/wiki/Tutorial#3-python-integration-for-embedding-and-plots).
Refer to wrapper function for further details on the inputs: https://github.com/ahof1704/gamma_bouts/blob/master/Pipeline/CBASS_L2_PlotEmbed.m
'''

import os
import numpy as np
import scprep
import matplotlib.pyplot as plt
import umap
from sklearn.decomposition import PCA
import scipy.io as sio
from scipy import stats
import matplotlib
from matplotlib import animation
import argparse
from IPython.display import HTML

def str2bool(v):
    if isinstance(v, bool):
       return v
    if v.lower() in ('yes', 'true', 't', 'y', '1'):
        return True
    elif v.lower() in ('no', 'false', 'f', 'n', '0'):
        return False
    else:
        raise argparse.ArgumentTypeError('Boolean value expected.')

parser = argparse.ArgumentParser()
parser.add_argument('--tmpPath', type=str)
parser.add_argument('--outPath', type=str)
parser.add_argument('--n_components', type=int)
parser.add_argument('--method', type=str)
parser.add_argument('--experiment', type=str)
parser.add_argument('--label_name', type=str)
parser.add_argument('--add_legend', type=str2bool) 
parser.add_argument('--fontsize', type=int) 
parser.add_argument('--discrete', type=str2bool) 
parser.add_argument('--category', type=int)
parser.add_argument('--format_img', type=str)
parser.add_argument('--rotate3D', type=str2bool)
args = parser.parse_args()

matplotlib.use('Agg')

mat_vars = sio.loadmat(os.path.join(args.tmpPath,'tmp_vars_' + args.experiment + '_' + args.label_name + '.mat'))
emb_data = sio.loadmat(os.path.join(args.outPath,'emb_data_' + args.method + '_' + args.experiment + '.mat'))['emb_data']

if '20Clusters' in args.label_name:
    cmap = 'tab20'
else:
    cmap = 'viridis'

labels = mat_vars['labels'][0]
# time_stamps = mat_vars['time_stamps']

# labels = bl1Run[time_stamps]
# Not sure if we will want to keep it or not
# if args.category >= 0:
#     category=args.category #1-> Running, 0-> Not running
#     print('category: {}'.format(category))
#     idx_run = np.where(labels[time_stamps]==category)
#     idx_run = idx_run[0]
# else:
#     idx_run = range(0,emb_data.shape[0]) #Just frame number
    
if args.n_components == 3:    
	if args.rotate3D == True:
		scprep.plot.rotate_scatter3d(emb_data, c=labels, filename = args.outPath + '/plot_embedding_' + str(args.n_components) + 'D_' + args.method + '_' + args.experiment + '_' + args.label_name + '.gif', 
		                             figsize=(10, 8), discrete = args.discrete, fontsize = args.fontsize, legend = args.add_legend, colorbar = args.add_legend, cmap=cmap, s=10, alpha=0.1)
	else:
		if args.format_img == 'eps':
			scprep.plot.scatter3d( emb_data, c=labels, filename = args.outPath + '/plot_embedding_' + str(args.n_components) + 'D_' + args.method + '_' + args.experiment + '_' + args.label_name  + '.' + args.format_img, 
				figsize=(10, 8), discrete = args.discrete, fontsize = args.fontsize, legend = args.add_legend, colorbar = args.add_legend, cmap=cmap, s=10)
		else:
			scprep.plot.scatter3d( emb_data, c=labels, filename = args.outPath + '/plot_embedding_' + str(args.n_components) + 'D_' + args.method + '_' + args.experiment + '_' + args.label_name  + '.' + args.format_img,
				dpi = 1000, discrete = args.discrete, colorbar = args.add_legend, fontsize = args.fontsize, figsize=(10, 8), cmap=cmap, s=10, alpha=0.1)
else:
	if args.format_img == 'eps':
		scprep.plot.scatter2d(emb_data, c=labels, filename = args.outPath + '/plot_embedding_' + str(args.n_components) + 'D_' + args.method + '_' + args.experiment + '_' + args.label_name + '.' + args.format_img, 
		                             figsize=(10, 8), discrete = args.discrete, fontsize = args.fontsize, legend = args.add_legend, colorbar = args.add_legend, cmap=cmap, s=10)
	else:
		scprep.plot.scatter2d(emb_data, c=labels, filename = args.outPath + '/plot_embedding_' + str(args.n_components) + 'D_' + args.method + '_' + args.experiment + '_' + args.label_name + '.' + args.format_img, 
		                             dpi = 1000, discrete = args.discrete, fontsize = args.fontsize, legend = args.add_legend, colorbar = args.add_legend, figsize=(10, 8), cmap=cmap, s=10, alpha=0.1)

plt.close('all')

File: matlab/Pipeline/CBASS_L2_PlotEmbed.m
function [status, commandOut] = CBASS_L2_PlotEmbed(sTROUGH, chTmpPath, chOutPath, chExperiment, chLabelTag, ...
    chMethod, chDataFormat, blZScore, inN_Component, db1Label, chFormatImg, chRotate3D, chAddLegend, inFontSize, chDiscrete)
% Wrapper function for matlab calling the python function CBASS_L2_PlotEmbed.py
% We use this function for to plot the 
%
% Input -------------------------------------------------------------------
%
% sTROUGH:          a structure containing the following fields:
%                   -.db1FilterBand an (1 x 2) array describing the frequency
%                   band of interest i.e. [30 80] for 30 to 80 Hz.
%                   -.db2Trough  a (2 * channel x trough) matrix containing the
%                   hilbert transform of each channel of sREC.db2LFP filtered
%                   in the band defined in sTROUGH.db1FilterBand at the trough
%                   of the filtered signal in the reference channel inRefChan
%                   -.in1Index the indices of the trough in sREC.db2LFP
% chTmpPath:        a character array indicating the temporary directory where
%                   files used by the embedding function will be saved.
% chOutPath:        a character array indicating the path to the output folder, 
%                   in which the embedded data will be saved.
% chExperiment:     character array indicating the name of the experiment and any
%                   other information that should be contained in the name of 
%                   saved file.
% chLabelTag        character array dedicated to labelling embedding plots using
%                   different plotting settings.
% chMethod:         character array specifying the method used for embedding. 
%                   The options are UMAP, PHATE and PCA. The default is 'UMAP'.      
% chDataFormat:     (optional) a character array specifying the format of
%                   the hilbert transforms output. Can be 'complex' or
%                   'polar'. Default is 'complex'.
% blZScore:         logical specifying if trought data is to be zscored for 
%                   k-means partitioning. Default is True.
% inN_Component:    (optional) a number between one and three specifying the 
%                   dimensionality of the embedded space. The default is a '3'.
% db1Labels:        array containing the labels associated to the data samples being 
%                   plotted. It is used to defined the color code of the plot. 
%                   The default is coloring samples according to their odering 
%                   (eg, timestamps)
% chFormatImg:      character array specifying the format of the output image.
%                   The options are 'png', 'eps, and 'gif'. The default is 'png'.
% chRotate3D:       character array specifying if the 3D plot will be saved as rotating 
%                   gif image or single image (png or eps). The default is 'False'.
% chAddLegend:      character array to hide ('False') or include ('True') the legend
%                   to the figure. The default is 'True'.
% inFontSize:       integer specifying fontsize of the plot. The default is 10.
% chDiscrete:       string specifying the settings for the legend. Use 'True' for discrete colorbar 
%                   and 'False' is for continuous. The default is 'True'.

% Output ------------------------------------------------------------------
% status            command exit status. When the command is successful, status is 0. 
%                   Otherwise, status is a nonzero integer.
% commandOut        Output of the operating system command, returned as a character vector.

% Checks argument
narginchk(5, 15)
if ~exist('chMethod', 'var'),       chMethod = 'umap';      elseif isempty(chMethod),       chMethod = 'umap'; end
if ~exist('chDataFormat', 'var'),   chDataFormat = 'polar'; elseif isempty(chDataFormat),   chDataFormat = 'polar'; end
if ~exist('blZScore', 'var'),       blZScore = true;        elseif isempty(blZScore),       blZScore = true; end
if ~exist('inN_Component', 'var'),  inN_Component = 3;      elseif isempty(inN_Component),  inN_Component = 3; end
if ~exist('chFormatImg', 'var'),    chFormatImg = 'png';    elseif isempty(chFormatImg),    chFormatImg = 'png'; end
if ~exist('chRotate3D', 'var'),     chRotate3D = 'False';   elseif isempty(chRotate3D),     chRotate3D = 'False'; end
if ~exist('chAddLegend', 'var'),    chAddLegend = 'True';   elseif isempty(chAddLegend),    chAddLegend = 'True'; end
if ~exist('inFontSize', 'var'),     inFontSize = 10;        elseif isempty(inFontSize),     inFontSize = 10; end
if ~exist('chDiscrete', 'var'),     chDiscrete = 'True';    elseif isempty(chDiscrete),     chDiscrete = 'True'; end
if ~exist('db1Label', 'var'),       db1Label = []; end

% Checks the input
if ~ismember(chMethod, {'umap', 'phate', 'pca'})
    fprintf('%s is not a valid method set to umap', chMethod); chMethod = 'umap'; 
end
if ~ismember(chDataFormat, {'polar', 'complex'})
    fprintf('%s is not a valid method set to umap', chDataFormat); 
    chDataFormat = 'complex'; 
end

% Export variables to temp for test
data            = sTROUGH.db2Trough;
    % Convert data to the wanted representation if wanted (the complex
    % trough can be represented in complex or polar coordinates)
if strcmp(sTROUGH.chDataFormat, 'complex') && strcmp(chDataFormat, 'polar')
    inNChan = size(data, 2)/2;
    data    = [abs(data(:, 1:inNChan) + 1i * data(:, inNChan + 1: end)), ...
        angle(data(:, 1:inNChan) + 1i * data(:, inNChan + 1: end))]; 
elseif strcmp(sTROUGH.chDataFormat, 'polar') && strcmp(chDataFormat, 'complex')
    inNChan     = size(data, 2)/2;
    db2DatCmplx = data(:, 1:inNChan) .* exp(1i * data(:, inNChan + 1:end));
    data        = [real(db2DatCmplx) imag(db2DatCmplx)]; 
end
if blZScore, data = zscore(data); end %zscore the data if required
time_stamps     = sTROUGH.in1Index - 1; %Time stamps of interest adjusted for python indexing
% Sets the labels
if isempty(db1Label), labels = 1:size(data,1); % Just the frames for now
else, labels = db1Label; end
method_embed    = chMethod; % or 'umap' default: phate
n_components    = inN_Component;
% chFormatImg     = 'png'
% chRotate3D        = 'False'
% chAddLegend      = 'False'
category        = -1; %To select running, for example, "-1" will just use frame numbers for labeling

save(fullfile(chTmpPath,['tmp_vars_' chExperiment '_' chLabelTag '.mat']), 'data', 'time_stamps', 'labels')

% Run this if you just want to plot the embedding
tic
disp(['Plotting the embedded data...'])
commandStr = "python ../Pipeline/CBASS_L2_PlotEmbed.py " + ... 
    " --tmpPath " + chTmpPath + ...
    " --experiment " + chExperiment + ...
    " --n_components " + num2str(n_components) + ...
    " --method " + method_embed + ...
    " --category " + num2str(category) + ...
    " --label_name " + chLabelTag + ...
    " --add_legend " + chAddLegend + ...
    " --format_img " + chFormatImg + ...
    " --fontsize " + inFontSize + ...
    " --discrete " + chDiscrete + ...
    " --rotate3D " + chRotate3D + ...
    " --outPath " + chOutPath + "" ;
% [status, commandOut] = system(sprintf(commandStr));
[status, commandOut] = system(sprintf('conda activate gammaBouts_env && %s', commandStr));
commandStr
if status ~= 0 % For some reason, the cluster only works with 'source'. I am adding this as an option.
    [status, commandOut] = system(sprintf('source activate gammaBouts_env && %s', commandStr));
end
toc